### ISSUE TITLE
AT+CREG=2 reponse refused

### ISSUE THREAD

**Issue by fulup-bzh** [gammu/gammu#91]
I'm trying to run Gammu with a GenProE10/Watcom SMS modem. 

Find out two issues 
* PIN control does not wait when modem to be ready. Easy to bypass by removing pin.
* AT+CREG=2 Watcom modem respond first with OK and only after with +CREG: 5,"25FF","825E. Gammu does not expect this behaviour, it retries for ever AT+CREG.

Question: Is is possible by configuration or by hacking the code to solve the AT++CREG=2 issue ? 

Note: I'm running 1.34 compiled from Gammu tar file on Ubuntu 14.04, with a MySql backend. GenProE10 Watcom AT supported commands is available at http://www.ercogener.com/DOC/support/Documents_Techniques/EG_GenProxxe_988_CL_003_UK.zip

Here after the trace for CREG, if I understand correctly. Gammu concider that it did not get a valid answer to AT+CREG and restart the command. at new command and this for ever.

```
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 1 "AT+CREG=2"
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 2 "OK"
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: Checking line: OK
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: AT reply state: 1
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: RECEIVED frametype 0x00/length 0x10/16
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 41A|54T|2B+|43C|52R|45E|47G|3D=|322|0D |0D |0A |4FO|4BK|0D |0A  AT+CREG=2...OK..
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: Enable full packet network info
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: SENDING frametype 0x00/length 0x0B/11
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 41A|54T|2B+|43C|47G|52R|45E|47G|3D=|322|0D                      AT+CGREG=2.     
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 1 "+CREG: 5,"25FF","825E""
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 2 "AT+CGREG=2"
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 3 "OK"
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: Checking line: OK
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: AT reply state: 1
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: RECEIVED frametype 0x00/length 0x29/41
Mon 2015/02/09 18:59:39 gammu-smsd[15464]: gammu: 2B+|43C|52R|45E|47G|3A:|20 |355|2C,|22"|322|355|46F|46F|22"|

```

For Pin code +CMS ERROR: 515 is not supported, and Gummy loop on AT-CPMS? when error 515 means operation in progress please wait. Obviously removing PIN solve this issue.
```
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: 1 "AT+CPIN?"
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: 2 "+CPIN: READY"
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: Checking line: +CPIN: READY
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: AT reply state: 1
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: RECEIVED frametype 0x00/length 0x19/25
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: 41A|54T|2B+|43C|50P|49I|4EN|3F?|0D |0D |0A |2B+|43C|50P|49I|4EN AT+CPIN?...+CPIN
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: 3A:|20 |52R|45E|41A|44D|59Y|0D |0A                              : READY..      
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: Parsing +CPIN: READY with +CPIN: @r
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: Grabbed string from reply: "READY" (parsed 5 bytes)
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: Parsed raw string "READY"
Mon 2015/02/09 16:41:11 gammu-smsd[6345]: gammu: Security status received - nothing to enter
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: Leaving GSM_GetSecurityStatus
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: Entering GSM_GetSMSStatus
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: Getting available SMS memories
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: SENDING frametype 0x00/length 0x0A/10
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: 41A|54T|2B+|43C|50P|4DM|53S|3D=|3F?|0D                          AT+CPMS=?.     
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: 1 "AT+CPMS=?"
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: 2 "+CMS ERROR: 515"
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: Checking line: +CMS ERROR: 515
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: AT reply state: 5
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: RECEIVED frametype 0x00/length 0x1D/29
Mon 2015/02/09 16:41:12 gammu-smsd[6345]: gammu: 41A|54T|2B+|43C|50P|4DM|53S|3D=|3F?|0D |0D |0A |2B+|43C|4DM|53S AT+CPMS=?...+C
    
```

**Comment by nijel** [comment]
This might be actually similar issue as #88, can you check whether the fix from there will help you?

**Comment by fulup-bzh** [comment]
I will download git version and try it. Nevertheless as I already hard cut many status request for my code It may not change much for me.

Since my last post I did intensive debug session. The critical missing missing feature is the lack of support for "+CMS ERROR: 515" [Busy, please retry later].

Because they is no option to return a "Retry later" after a CMS error. The only solution we have, is to add 'long sleep" to avoid cases where the modem may not be ready before Gammu send its next command. 

If the modem had an option to notify Gammu to wait 1s before retrying the same command it would solve many issues.

**Comment by fulup-bzh** [comment]
I checked #88, installed the latest commit and did some intensive debug session. Unfortunately I fail to understand state machine logic well enough to find a work around to  "AT+CREG=2" error. I nevertheless probably have my finder on the error.

When "AT+CREG=2" is sent and receive because of the echo mode. Gammu enter in ATGEN_ReplyGetNetworkLAC_CID where the "OK" answer is detected. I tried to change return code and replace it with ERR_NEEDANOTHERANSWER with no success.

At this point we would like Gammu to ignore the "OK" and force it to reenter in the message waiting loop to wait for next coming line that is the line we're waiting for.

What ever return code I try. Gammu resent the command  "AT+CREG=2" before it get the answer. 
```
	switch (s->Phone.Data.Priv.ATGEN.ReplyState) {
	case AT_Reply_OK:
        smprintf(s, "Fulup got OK\n");
     	return ERR_NEEDANOTHERANSWER;

		break;
	case AT_Reply_CMSError:
	        return ATGEN_HandleCMSError(s);
	case AT_Reply_CMEError:
		return ATGEN_HandleCMEError(s);
	default:
		return ERR_UNKNOWNRESPONSE;
	}
```

Note: on last commit. I saw that you add a retry for  "+CMS ERROR: 515" for "Retry Later" but today my modem refuse to enter in timeout mode, also I'm not able to test it :(
